hal/include/HAL/Encoder.h

```diff
@@ -9,26 +9,57 @@
 
 #include <stdint.h>
 
+#include "HAL/Handles.h"
+
 extern "C" {
-void* initializeEncoder(uint8_t port_a_module, uint32_t port_a_pin,
-                        bool port_a_analog_trigger, uint8_t port_b_module,
-                        uint32_t port_b_pin, bool port_b_analog_trigger,
-                        bool reverseDirection, int32_t* index,
-                        int32_t* status);  // TODO: fix routing
-void freeEncoder(void* encoder_pointer, int32_t* status);
-void resetEncoder(void* encoder_pointer, int32_t* status);
-int32_t getEncoder(void* encoder_pointer, int32_t* status);  // Raw value
-double getEncoderPeriod(void* encoder_pointer, int32_t* status);
-void setEncoderMaxPeriod(void* encoder_pointer, double maxPeriod,
-                         int32_t* status);
-bool getEncoderStopped(void* encoder_pointer, int32_t* status);
-bool getEncoderDirection(void* encoder_pointer, int32_t* status);
-void setEncoderReverseDirection(void* encoder_pointer, bool reverseDirection,
+enum EncoderIndexingType {
+  HAL_kResetWhileHigh,
+  HAL_kResetWhileLow,
+  HAL_kResetOnFallingEdge,
+  HAL_kResetOnRisingEdge
+};
+enum EncoderEncodingType { HAL_Encoder_k1X, HAL_Encoder_k2X, HAL_Encoder_k4X };
+
+HalEncoderHandle initializeEncoder(
+    uint8_t port_a_module, uint32_t port_a_pin, bool port_a_analog_trigger,
```

Comment:
We could take a Handle plus a mode. But then we have to change the DigitalSource Interface upstream, and change all the classes that currently depended on that, along with changing interrupts at the HAL level as well. I wanted to do that after this got merged cause it was already really big. The reason we need a mode and a non typed handle is that can be uses with AnalogTriggers, which require both data from the AnalogTrigger handle and the mode type, whereas Digital only requires the Digital Index.